src/dlr_relayvm.cc

```diff
@@ -0,0 +1,290 @@
+  std::string input_name(name);
+  for (auto i = 0; i < num_inputs_; i++) {
+    if (input_name == input_names_[i]) {
+      return i;
+    }
+  }
+  throw dmlc::Error("Invalid input node name!");
+}
+
+DLDataType RelayVMModel::GetInputDLDataType(int index) {
+  auto input_type = input_types_[index];
+  DLDataType dtype;
+  dtype.lanes = 1;
+  if (input_type == "uint8") {
+    dtype.code = kDLUInt;
+    dtype.bits = 8;
+  } else if (input_type == "int8") {
+    dtype.code = kDLInt;
+    dtype.bits = 8;
+  } else if (input_type == "float32") {
+    dtype.code = kDLFloat;
+    dtype.bits = 32;
+  } else if (input_type == "float64") {
+    dtype.code = kDLBfloat;
+    dtype.bits = 64;
+  } else {
+    throw dmlc::Error("Unknown input dtype!");
+  }
+  return dtype;
+}
+
+void RelayVMModel::SetInput(const char *name, const int64_t *shape, void *input,
+                            int dim) {
+  int index = GetInputIndex(name);
+  DLDataType dtype = GetInputDLDataType(index);
+  DLTensor input_tensor;
+  input_tensor.data = input;
+  input_tensor.ctx = ctx_;
+  input_tensor.ndim = dim;
+  input_tensor.shape = const_cast<int64_t *>(shape);
+  input_tensor.strides = nullptr;
+  input_tensor.byte_offset = 0;
+  input_tensor.dtype = dtype;
+  std::vector<int64_t> arr_shape(shape, shape + dim);
+
+  tvm::runtime::NDArray input_arr =
+      tvm::runtime::NDArray::Empty(arr_shape, dtype, ctx_);
+  input_arr.CopyFrom(&input_tensor);
+  inputs_[index] = input_arr;
+}
+
+void RelayVMModel::UpdateInputs() {
+  const int kNumArgs = GetNumInputs() + 1;
+  TVMValue values[kNumArgs];
+  int type_codes[kNumArgs];
+  auto arg_setter = tvm::runtime::TVMArgsSetter(values, type_codes);
+  arg_setter(0, ENTRY_FUNCTION);
+  for (int i = 0; i < inputs_.size(); i++) {
+    arg_setter(i + 1, inputs_[i]);
+  }
+
+  tvm::runtime::PackedFunc set_input = vm_module_->GetFunction("set_input");
+  tvm::runtime::TVMRetValue rv;
+  set_input.CallPacked(tvm::runtime::TVMArgs(values, type_codes, kNumArgs),
+                       &rv);
+}
+
+void RelayVMModel::Run() {
+  // Invoke inference
+  UpdateInputs();
+  tvm::runtime::PackedFunc invoke = vm_module_->GetFunction("invoke");
+  output_ref_ = invoke(ENTRY_FUNCTION);
+  UpdateOutputs();
+}
+
+void RelayVMModel::UpdateOutputs() {
+  outputs_.resize(num_outputs_);
+  if (output_ref_->IsInstance<tvm::runtime::ADTObj>()) {
+    auto adt = tvm::runtime::Downcast<tvm::runtime::ADT>(output_ref_);
+    for (int i = 0; i < adt.size(); i++) {
+      outputs_[i] = tvm::runtime::Downcast<tvm::runtime::NDArray>(adt[i]);
+    }
+  } else if (output_ref_->IsInstance<tvm::runtime::NDArray::ContainerType>()) {
+    outputs_[0] = tvm::runtime::Downcast<tvm::runtime::NDArray>(output_ref_);
+  } else {
+    throw dmlc::Error("Invalid output_ref format!");
+  }
+}
+
+void RelayVMModel::GetOutput(int index, void *output) {
+  CHECK_LT(index, num_outputs_) << "Output index is out of range.";
+  auto out_array = outputs_[index];
+  DLTensor output_tensor;
```

Comment:
You can keep this for now. I think a simpler way could be something like `std::memcpy(output, out_array->data, GetDataSize(*(out_array->operator())`

Reply:
👍 I will clean this up later along with tvm in the refactor PR

Reply:
ah, we actually need tensor copy here because we may need to copy data from different devices.